homeassistant/runner.py

```diff
@@ -116,6 +103,88 @@ def new_event_loop(self) -> asyncio.AbstractEventLoop:
         return loop
 
 
+class HassEventLoop(uvloop.Loop):
+    """Loop which exposes timer handles."""
+
+    def __init__(self) -> None:
+        """Initialize the event loop."""
+        super().__init__()
+        self._scheduled: set[asyncio.TimerHandle] = set()
+
+    def _prune_cancellable_timers(self) -> None:
+        temp_timers = self._scheduled.copy()
+        for handle in temp_timers:
+            if handle.cancelled() or handle.when() < self.time():
+                self._scheduled.remove(handle)
+
+    def _handle_cancellable_timer(
+        self, timer: asyncio.TimerHandle, *args: ParamSpecArgs
+    ) -> None:
+        if (
+            timer not in self._scheduled
+            and args is not None
+            and len(args) > 0
+            and isinstance(args[0], HassJob)
+            and args[0].cancel_on_shutdown
+        ):
+            self._scheduled.add(timer)
+
+    def call_at(
+        self,
+        when: float,
+        cb: Callable[[Any, Any], Any],
+        *args: Any,
+        context: Any | None = None,
+    ) -> asyncio.TimerHandle:
+        # pylint: disable=arguments-differ
+        """Call callback at a future time.
+
+        Overridden from base class to track cancellable timers
+        """
+        self._prune_cancellable_timers()
+        timer = super().call_at(when, cb, *args, context=context)
+        self._handle_cancellable_timer(timer, *args)
+        return timer
+
+    def call_later(
+        self,
+        delay: float,
+        cb: Callable[[Any, Any], Any],
+        *args: ParamSpecArgs,
+        context: Any | None = None,
+    ) -> asyncio.TimerHandle:
+        # pylint: disable=arguments-differ
+        """Call coroutine later.
+
+        Overridden from base class to track cancellable timers
+        """
+        self._prune_cancellable_timers()
+        timer = super().call_later(delay, cb, *args, context=context)
+        self._handle_cancellable_timer(timer, *args)
+        return timer
+
+    def create_future(self) -> asyncio.Future[Any]:
```

Comment:
There are quite many places in the project where we call `asyncio.Future` directly instead of using `loop.create_future` which is recommended to cater for custom event loops. This may need to be solved in this PR too.

Reply:
I thought set_event_loop_policy() takes care of that?

The only reason I overrode the method was to periodically prune the HassJob cancellable timers which I need to manually keep track of.

I figured that was the best place to do it since it won’t effect the event loop too much and HA creates lots of tasks.

Reply:
Why do you need to prune the timers?

Reply:
I don't see that there is a way to access `loop._timers` in uvloop.

It would be nicer if there was a public method to do that as it would avoid the complexity here.  I don't know how willing they would be to add something like that

Reply:
https://github.com/bdraco/uvloop/commit/5f8e84e22cc266b25db595e6dfc71d85cfa0b0b2

If you could get something like that merged to uvloop you could avoid the subclassing

Reply:
`add_callback_done` might be an alternative as well

Reply:
Also I really don’t like that we need to access scheduled in our current code base so if there is another option to refactor HA to avoid it that would also be a welcome change (if I had a suggestion to do that I think we already would have )

Reply:
There is a private attribute that acts similarly to _scheduled in the uvloop code. I just don’t know how to access it from Python.

Reply:
> Why do you need to prune the timers?
> 
> Edit: you need to access `loop._scheduled`

Yeah, if I knew how to do that, all of this method overriding would be unnecessary. I’m working around the fact I can’t access it by just keeping a list of cancellable timers. 

I don’t know how to access an attribute coded in C in Python.

Reply:
> Also I really don’t like that we need to access `_scheduled` in our current code base so if there is another option to refactor HA to avoid it that would also be a welcome change (if I had a suggestion to do that I think we already would have )

Add it to the Hass object?

Reply:
> [bdraco/uvloop@5f8e84e](https://github.com/bdraco/uvloop/commit/5f8e84e22cc266b25db595e6dfc71d85cfa0b0b2)
> 
> If you could get something like that merged to uvloop you could avoid the subclassing

Well, I originally tried to access _timers directly, but I couldn't figure out how to as it is in C.